Drop Python 3.8 from CI matrix.

danielballan opened the pull request:
Python 3.8 reached its end-of-life (per [Python upstream](https://devguide.python.org/versions/)) on October 7, a week ago today.

danielballan commented:
There is ongoing discussion among the Technical Steering Committee about which versions of Python Bluesky projects should support:

- Follow CPython upstream?
- Follow SPEC 0?
- More aggressive?

and the answer may be different for different projects. However, even while we wait for that discussion to conclude, I think we can say confidently that Bluesky cannot support versions of Python that are no longer supported by upstream.

callumforrester reviewed (changes requested):
@danielballan Could you also drop it from `pyproject.toml` so that PyPi reflects this change?

danielballan commented:
Good catch! Done.

prjemian commented:
Note that Py3.13 is [now supported](https://devguide.python.org/versions/).  New work needed for this throughout the framework.

tacaswell commented:
> New work needed for this throughout the framework.

The biggest problem I expect is waiting for things with rust extensions to release new versions that pin PyO3 high enough to build 3.13 extensions.  I've been doing my (limited) dev work in a 3.13 environment for a few months now and not had any major surprises.

evalott100 commented:
This needs bumping too

https://github.com/bluesky/bluesky/blob/3bce96b85558993499174be7eebcd09f270a9271/pyproject.toml#L35

mrakitin reviewed (approved):
Looks good to me. Let's move on to avoid failures in other PRs for the outdated version of Python.

danielballan commented:
Rebased and addressed @tangkong's comments. Only remaining instances of `3.8` in the codebase:

```
$ git grep "3\.8 "
docs/api_changes.rst:Added support for Python 3.8 and the following for forward-compatibility with
src/bluesky/tests/test_callbacks.py:|        43  |  04:17:23.8  |          0.85  |          0.56  |
```

danielballan commented:
The linter is failing on unrelated complaints, looks like increased strictness.

tacaswell commented:
Merged over the lint failure, looks like ruff got stricter about using base types as generics instead of the special ones from typing (e.g. `tuple` vs `Tuple`).

danielballan commented:
I expect that is related to Python 3.8 going EOL, as Python 3.8 was the last version that did not support using generics as types.
